### PR TITLE
Add date filter to request logs

### PR DESCRIPTION
**Pull request by arcticfly**
With this change, we can now filter request logs by date.

### Changes
* Add `text | date` type field to filter option
* Split filter into TextFilter and DateFilter
* Add `LAST 15M`, `LAST 24H`, `LAST 7D`, `BEFORE`, `AFTER` and `RANGE` date comparators
* Add `"Sent At"` as a filterable field for request logs

Sent At filters:
Screenshot 2023-10-30 at 1 15 13 AM

### CODE REVIEW

File: app/src/components/Filters/AddFilterButton.tsx [hunk 1]
```diff
@@ -1,8 +1,10 @@
 import { Button, HStack, Icon, Text } from "@chakra-ui/react";
 import { BsPlus } from "react-icons/bs";
-import { comparators, useFilters } from "./useFilters";
+import { useFilters } from "./useFilters";
+import { comparators } from "~/types/shared.types";
+import { type FilterOptionType } from "./types";
 
-const AddFilterButton = ({ filterOptions }: { filterOptions: string[] }) => {
+const AddFilterButton = ({ filterOptions }: { filterOptions: FilterOptionType[] }) => {
```
Comment (corbt) [inline app/src/components/Filters/AddFilterButton.tsx]: Minor, but we could define a helper type like follows:

`type AtLeastOne<T> = [T, ...T[]];`

And then change the type here to `filterOptions: AtLeastOne<FilterOptionType>`.

That way TS would enforce there's at least one filter option passes, so you wouldn't have to cast the `as string` below.

File: app/src/components/Filters/Filter/DateFilter.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      updateDate("");
+    }
+  }, [valueIsEmpty, updateDate]);
+
+  return (
+    <HStack>
+      <SelectFieldDropdown filterOptions={filterOptions} filter={filter} />
+      <SelectComparatorDropdown filter={filter} filterType="date" />
+      <HStack>
+        {!["LAST 15M", "LAST 24H", "LAST 7D"].includes(filter.comparator) && (
```
Comment (corbt) [inline app/src/components/Filters/Filter/DateFilter.tsx:71]: This is a nit but can we update the way we display these at least to be title case (eg. "Last 15M", "Before")? They kinda feel like they're shouting at me. 😁

File: app/src/server/utils/constructLoggedCallFiltersQuery.ts [hunk 3]
```diff
@@ -31,8 +29,41 @@ const comparatorToSqlExpression = (comparator: (typeof comparators)[number], val
   };
 };
 
-export const constructFiltersQuery = (
-  filters: z.infer<typeof logFiltersSchema>,
+export const dateComparatorToSqlExpression = (
+  comparator: (typeof comparators)[number],
+  value: number[],
```
Comment (corbt) [inline app/src/server/utils/constructLoggedCallFiltersQuery.ts]: It loks like this is always supposed to be of length 2? If so you can define it as a tuple eg. `value: [number, number]`.

File: app/src/server/utils/constructLoggedCallFiltersQuery.ts [hunk 4]
```diff
@@ -46,20 +77,41 @@ export const constructFiltersQuery = (
     .where((eb) => {
       const wheres: Expression<SqlBool>[] = [eb("lc.projectId", "=", projectId)];
 
+      console.log("filters", filters);
```
Comment (corbt) [inline app/src/server/utils/constructLoggedCallFiltersQuery.ts]: rm

**Review by corbt — APPROVED** [review]
Few small comments but looks good!

File: app/src/components/Filters/AddFilterButton.tsx (hunk 1 above)
Comment (arcticfly) [reply app/src/components/Filters/AddFilterButton.tsx]: Yes, this does save some casting (unfortunately it adds an `unknown` casting in `Filters.tsx`, but at least the reasoning behind the cast is explained by its surrounding code).

**Pull request merged**